presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
Also what about sorting when `WindowNode`s are separated with `ProjectNode`, like below?

```
sum over (partition by A order by C)
project()
avg over (partition by B)
project()
sum over (partition by A order by D)
```

This is why I thought that having a set of rules which will work together in order group all the `WindowNode`s together will open a bunch of opportunities for optimizer.

> Separating iterative optimizer rules into different optimizers defeats a main purpose of the iterative optimizer, and that is to allow the rules to work together to discover new plans that the developer might not have anticipated.

This is true, with a caveat it is not ;) Mostly because of `There are simply too many possible plans to explore even a small fraction of them.`. I expect to have some trade-offs like running `ReorderJoins` on its own etc. I could imagine a set of rules which could optimize window functions (it may also consists of column pruning rules, predicate pushdown, projection moving and merging rules rule and anything which could help your rules). Once `WindowNodes` are optimized (like nodes are merged or sorted) then subsequent iterative optimizers (or other optimizer stages) should not break this. `WindowNode` optimizations are not cost-based (do you expect a change here?) so the search space will not be so huge in such stage.